appsrc/dcfile/dckey.man
.TH DCKEY 1 "19 November 2008" "DICOM PS3" "DICOM PS3 - Extract attribute values"
.SH NAME
dckey \- ACR/NEMA DICOM PS3 ... Extract attribute values
.SH SYNOPSIS
.HP 10
.B dckey
[
.B \-ignorereaderrors
]
[
.B \-v|verbose
]
[
.B \-filename
]
[
.B \-describe
]
[
.B \-brief
]
[
.B \-d|decimal
]
[
.B \-key|k " elementname|(0xgggg,0xeeee) ] ..."
.so man1/optin.so
.SH DESCRIPTION
.LP
.B dckey
reads the named dicom input file and displays the values of the selected attributes.
.LP
Binary attributes are written in hexadecimal with a preceding
"0x" and numeric string attributes are written in decimal as they are encoded,
unless the \-decimal option is specified.
.SH OPTIONS
The attribute values, description and verbose output go to standard error.
.PP
The basic input switches are described in dcintro(1). Options specific to this program are:
.TP
.B \-describe
.RS
Display the tag, name, value representation and value length of the attribute.
.RE
.TP
.B \-brief
.RS
Display the keyword of the attribute, and equals sign, and then the value, for use in
scripts that automatically extract multiple values in the same dckey invocation. Implies \-noerror.
.RE
.TP
.B \-ignorereaderrors
.RS
Attempt to extract key values even if errors encountered whilst parsing DICOM file.
.RE
.TP
.B \-noerror
.RS
Do not report an error if the requested attribute was not found, just an empty value or line. Implied by \-brief.
.RE
.TP
.B \-d|decimal
.RS
Display the first (or only) value of a numeric attribute as a decimal value.
.RE
.TP
.B \-k|key " elementname|(0xgggg,0xeeee)"
.RS
Add a key to the list of attributes to be extracted. The attribute may also be specified
as a hexadecimal pair of the form (0xgggg,0xeeee), which also allows private attributes to be fetched.
.RE
.TP
.B \-filename
.RS
Show the name of the file supplied in the arguments; a hyphen will be reported if no filename was supplied.
.RE
.SH ENVIRONMENT
.LP
\ 
.SH EXAMPLES
.LP
.RE
\ 
.RE
% dckey -key Rows -key Columns -key ImageLocation
.RE
        -describe test.dc3
.RE
(0x0028,0x0010) US Rows    VR=<US> VL=<0x0002> [0x0100] 
.RE
(0x0028,0x0011) US Columns VR=<US> VL=<0x0002> [0x0100] 
.RE
Error - Not found - (0x0028,0x0200) US Image Location 
.RE
\ 
.RE
% dckey -k Rows -k ImagePositionPatient
.RE
      -k StudyDate -k StudyTime
.RE
      -k SOPClassUID test.dc3
.RE
1.2.840.10008.5.1.4.1.1.2
.RE
1996.04.17
.RE
19:38:00
.RE
45.000000\\-120.000000\\120.000000 
.RE
0x0200
\ 
.RE
% dckey -d -k Rows test.dc3
.RE
512
\ 
.RE
% dckey -brief -k Rows -k Columns test.dc3
.RE
Rows=0x200
.RE
Columns=0x200
.SH FILES
.LP
\ 
.SH SEE ALSO
.BR dcintro(1), dctable(1)
.SH AUTHOR
Copyright (C) 1993-2010. David A. Clunie DBA PixelMed Publishing. All rights reserved.
.SH BUGS
.LP
Attempts to extract sequences or OB,OW attributes may fail nastily.
.LP
The order in which the attributes are extracted and written is not the same
order as they are specified on the command line, and should not be depended
on.
.LP
The \-describe and other options cannot be used at the same time.
.LP
The \-decimal option uses a double conversion before displaying the value and
there may be sign extension and precision issues with some values.
.LP
The reading of the attributes from the DICOM file will not proceed past the
top level Pixel Data attribute, in order to accelerate the speed of reading.



CHANGES
111219:	libsrc/src/dctool/attrmxrd.cc: check for HighBit being more than one less than BitsStored and correct BitsStored accordingly (otherwise will create zero pixel values on copy) (000408)

111213:	appsrc/dcfile/dcuncat.cc: add option to replace instance number starting with a specific number to allow scripts to extend a series with successive invocations

111212:	libsrc/standard/condn.tpl/module,strval/base.tpl,dx.tpl: improve animal versus human conditions to check SRT homo sapiens, and make view position and body part examined human versus animal specific

111212:	libsrc/standard/elmdict/other.tpl: add more Sound private elements

111117:	libsrc/standard/strval/base.tpl: add CSD for 99NCIAIM

111109:	appsrc/dcfile/dcuncat.cc: add options to replace instance number, frame increment pointer and frame increment pointer vectors, as well as add number of frames during unenhance based on SOP class

111106:	libsrc/standard/elmdict/other.tpl: add Vepro private elements

111031:	libsrc/standard/elmdict/other.tpl: add CTP private elements

111030:	libsrc/standard/elmdict/other.tpl: add Segami private elements

111029:	libsrc/standard/elmdict/dicom3.tpl: add (restore from trial) Observation UID (per CP 1147)

111028:	libsrc/standard/elmdict/other.tpl: add NeuroQuant private elements

111024:	libsrc/standard/elmdict/other.tpl: update fuji private attributes with descriptions

110918:	libsrc/standard/elmdict/diconde.tpl: update with new or modified data elements with cleanup from Patrick

110918:	libsrc/standard/elmdict/siemens.tpl: update some private elements based on Artis conformance statement

110916:	libsrc/standard/elmdict/gems.tpl: resolve inconsistencies in GE PET conformance statement VRs SS/SL/FL for private elements by choosing those most commonly seen in the field (still some inconsistencies between models)

110904:	libsrc/standard/elmdict/camtron.tpl: add private data elements encountered in DICOMDIR, as well as QCA private data elements from GEMS/Camtronics CRS conformance statement

110901:	libsrc/standard/sopcl.tpl,elmdict/dicom3.tpl: Sup 152 FT ophthalmic thickness map (dictionary and SOP class only; not IOD yet)

110822:	libsrc/standard/elmdict/dicom3.tpl: update data dictionary (only) for CPs for FT Aug 2011 WG 6 meeting

110714:	libsrc/standard/elmdict/dicom3.tpl: remove 'unified' from names of various attributes (per Sup 96 FT)

110716:	libsrc/standard/strval/base.tpl: add 99PMP CSD

110714:	libsrc/standard/elmdict/dicom3.tpl: exposure index etc are DS not FD (CP 1024)

110709:	libsrc/standard/elmdict/dicom3.tpl: change (0018,9430) from US to FL per Sup 139

110612:	libsrc/standard/elmdict/diconde.tpl: update for DICONDE eddy current standard

110612:	libsrc/standard/elmdict/dicondep.tpl: correct case of DICONDE DX (Dx) private owner for E2699

110609:	libsrc/standard/elmdict/dicom3.tpl: add Sup 151 FT data elements

110608:	libsrc/standard/elmdict/elscint.tpl: update Elscint private elements

110531:	libsrc/standard/elmdict/other.tpl: update Hitachi CT private elements

110531:	libsrc/standard/elmdict/siemens.tpl: update mammo private elements

110529:	libsrc/standard/condn.tpl,iodcomp/dx.tpl,module/base.tpl,dx.tpl,xaxrf.tpl,strval/dx.tpl,xaxrf.tpl: add breast tomosynthesis IOD, refactoring 3D XRay accordingly

110529:	libsrc/standard/elmdict/other.tpl: more private data elements for Hologic 

110528:	appsrc/dcfile/dciodvfy.cc: do not complain about SCOORD without child IMAGE when SELECTED FROM is by reference (000406)

110527:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: check that references to instances in content tree are in top level evidence sequences in SR (000405)

110526:	libsrc/standard/elmdict/gems.tpl: add names for CT iterative reconstruction private data elements

110522:	libsrc/src/dctool/attrothr.cc: do not throw assertion if computed length is equal (which may occur for single bit images)

110522:	libsrc/standard/iodcomp/base.tpl,libsrc/standard/condn.tpl: permit FoR Module in Segementation (and do CP to standard to allow may be present otherwise)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect SCOORD without child IMAGE (000403)

110522:	appsrc/dcfile/dciodvfy.cc,libsrc/src/locale/mesgtext.tpl: detect quotes in code meaning of measurement units (000404)

110522:	libsrc/src/dctool/attrtypv.cc,locale/mesgtext.tpl: permit PN with multiple carets in multiple component groups (000402)

110521:	libsrc/standard/elmdict/acuson.tpl,agfa.tpl,isg.tpl,other.tpl,siemens.tpl,strval/base.tpl: more private elements and coding scheme designators

110517:	libsrc/standard/module/base.tpl: correct type (1->2) and number of items (1->0-n) in SourceImageSequence in DerivationImageMacro

110517:	libsrc/standard/elmdict/siemens.tpl: yet more syngo private elements

110516:	libsrc/standard/cond.tpl: no longer require version for UCUM codes (CP 1060)

110516:	libsrc/standard/strval/base.tpl: add 99_MI private coding scheme designator

110516:	libsrc/include/dctool/attr.h,attrseq.h,generic/listsimp.h,src/dctool/attrseq.cc,attrvrfy.cc,standard/condn.tpl,module/softcopy.tpl,sr.tpl,support/module.awk:
		though VM for sequences number of items was being specified in modules, it was not being applied;
		apply it (000256), as well as make it work (since number of items was not being computed correctly
		in SequenceAttribute::verifyVM()); also add support for conditional check of VM, for example for
		GraphicData depending on Grahic Type (000401)

110516:	libsrc/standard/elmdict/toshiba.tpl: update CT private elements from conformance statement for V4.70 or later version

110514:	libsrc/standard/elmdict/siemens.tpl: add more syngo private elements, and update guessed VRs of existing ones with observed explicit VRs

110511:	libsrc/standard/elmdict/gems.tpl,toshiba.tpl: add more encountered explicit VR private elements

110511:	libsrc/standard/elmdict/other.tpl: update/add McKesson private elements from conformance statement


appsrc/dcfile/dckey.cc
#include "attrmxls.h"
#include "attrnew.h"
#include "attrval.h"
#include "attrseq.h"
#include "mesgtext.h"
#include "dcopt.h"
#include "dcstream.h"
#include "elmconst.h"

static const char *options_output_key[] = {
	"key",
	"k",
	0
};

// Same in dctable.cc

static void dumpCodeSequenceAsString(SequenceAttribute *as,TextOutputStream &log) {
	AttributeList **itemLists;
	int nItems = as->getLists(&itemLists);
	for (int i=0; i<nItems; ++i) {
		AttributeList *thisItemList = itemLists[i];
		if (thisItemList) {
			Attribute *aCodeValue = (*thisItemList)[TagFromName(CodeValue)];
			Attribute *aCodingSchemeDesignator = (*thisItemList)[TagFromName(CodingSchemeDesignator)];
			Attribute *aCodeMeaning = (*thisItemList)[TagFromName(CodeMeaning)];
			if (aCodeValue || aCodingSchemeDesignator || aCodeMeaning) {
				const char *vCodeValue = AttributeValue(aCodeValue,"");
				const char *vCodingSchemeDesignator = AttributeValue(aCodingSchemeDesignator,"");
				const char *vCodeMeaning = AttributeValue(aCodeMeaning,"");
				log << "('"
				    << vCodeValue
				    << "','"
				    << vCodingSchemeDesignator
				    << "','"
				    << vCodeMeaning
				    << "')"
				;
			}
		}
	}
}

int
main(int argc,char **argv)
{
	GetNamedOptions 	options(argc,argv);
	DicomInputOptions 	dicom_input_options(options);

	bool verbose=options.get("verbose") || options.get("v");
	bool showfilename=options.get("filename");

	bool noerror=options.get("noerror");
	bool ignorereaderrors=options.get("ignorereaderrors");
	bool brief=options.get("brief");
	bool describe=options.get("describe");
	bool decimal=options.get("decimal") || options.get("d");

	bool bad=false;

	// identical to dctable.cc; should refactor this, perhaps as AttributeTag::read() ? :(
	ElementDictionary optdict;
	AttributeList *keylist=0;
	const char *keyarg;
	while (options.get(options_output_key,keyarg)) {
		Assert(keyarg);
		Tag tag;
		if (!getAttributeTagFromStringHexForm(keyarg,tag)	// handle "-k (0xgggg,0xeeee)" case
		 && !optdict.getTag(keyarg,tag)) {			// handle "-k keyword" case
			bad=true;
			cerr << "-" << options_output_key[0] << ": "
			     << EMsgDC(UnrecognizedElement)
			     << " - \"" << keyarg << "\""
			     << endl;
		}
		else {
			const char *vr=optdict.getValueRepresentation(tag);
			if (!vr) vr="UN";
			Attribute *a=newAttribute(vr,tag);
			Assert(a);
			if (!keylist) {
				keylist=new AttributeList;
				Assert(keylist);
			}
			(*keylist)+=a;
		}
	}

	dicom_input_options.done();
	options.done();

	DicomInputOpenerFromOptions input_opener(
		options,dicom_input_options.filename,cin);

	cerr << dicom_input_options.errors();
	cerr << options.errors();
	cerr << input_opener.errors();

	if (!dicom_input_options.good()
	 || !options.good()
	 || !input_opener.good()
	 || !options
	 || bad) {
		cerr 	<< MMsgDC(Usage) << ": " << options.command()
			<< dicom_input_options.usage()
			<< " [-describe]"
			<< " [-noerror]"
			<< " [-ignorereaderrors]"
			<< " [-brief]"
			<< " [-decimal|d]"
			<< " [-key|k elementname|(0xgggg,0xeeee)]"
			<< " [-verbose|v]"
			<< " [-filename]"
			<< " [" << MMsgDC(InputFile) << "]"
			<< " <" << MMsgDC(InputFile)
			<< endl;
		return 1;
	}

	DicomInputStream din(*(istream *)input_opener,
		dicom_input_options.transfersyntaxuid,
		dicom_input_options.usemetaheader);

	ManagedAttributeList list;

	bool success=true;
	TextOutputStream  log(cerr);
	
	if (showfilename) {
		const char *filenameused = input_opener.getFilename();
		cerr << "Filename: \"" << (filenameused && strlen(filenameused) > 0 ? filenameused : "-") << "\"" << endl;
	}

	if (verbose) log << "******** While reading ... ********" << endl; 
	list.read(din,&log,verbose,0xffffffff,true,dicom_input_options.uselengthtoend,dicom_input_options.ignoreoutofordertags,dicom_input_options.useUSVRForLUTDataIfNotExplicit,
		true/*useStopAtTag*/,TagFromName(PixelData));

	if (!list.good()) {
		log << list.errors()
		    << EMsgDC(DatasetReadFailed) << endl;
		success=false;
	}
	if (success || ignorereaderrors) {
		ElementDictionary *dict=list.getDictionary();
		Assert(dict);
		AttributeListIterator listi(*keylist);
		while (!listi) {
			Attribute *ak=listi();
			Assert(ak);
			Tag tag=ak->getTag();
			Attribute *ad=list[tag];
			if (!describe && brief) {
				const char *keyword=0;
				if (dict) {
					keyword=dict->getKeyword(tag);	// not description; want without spaces
				}
				if (keyword) {
					log << keyword;
				}
				else {
					log << "(";
					writeZeroPaddedHexNumber(log,tag.getGroup(),4);
					log << ",";
					writeZeroPaddedHexNumber(log,tag.getElement(),4);
					log << ") ";
				}
				log << "=";
			}
			if (ad) {
				if (describe) {
					ad->write(log,dict);
				}
				else {
					if (decimal) {
						log << double(AttributeValue(ad,double(0)));
					}
					else {
						if (ad->isSequence()) {
							SequenceAttribute *as = (SequenceAttribute *)ad;
							dumpCodeSequenceAsString(as,log);
						}
						else {
							ad->writeData(log);
						}
					}
				}
			}
			else {
				if (!noerror && !brief) {
					log << EMsgDC(NotFound) << " - ";
					tag.write(log,dict);
				}
			}
			log << endl;
			++listi;
		}
	}

	return success ? 0 : 1;
}

